### PR TITLE
fix(sql): ASOF and LT JOINS no longer convert SYMBOL keys to STRING

### LINKED ISSUE #2976: Multiple LT JOINs error on SYMBOL column
### Describe the bug

on this table:
```
CREATE TABLE 'tests' (
  Ticker SYMBOL capacity 256 CACHE,
  H FLOAT,
  P_T FLOAT,
  timestamp TIMESTAMP
) timestamp (timestamp) PARTITION BY MONTH;
```
After inserting data, calling from webui:
`tests LT JOIN (tests LT JOIN (tests) ON (Ticker)) ON (Ticker)`
resulting Ticker type changes to STRING instead of SYMBOL on Ticker1
This causes further problems when adding more ON statements. In the following select:
`tests LT JOIN (tests LT JOIN (tests LT JOIN (tests)) ON (Ticker)) ON (Ticker)`
The following error message occurs:
`right side column 'Ticker1' is of unsupported type`
Which is the default error message when trying to LT JOIN ON STRING column

### To reproduce

1. Create the table as outlined in description
2. insert data
3. perform the following query:`tests LT JOIN (tests LT JOIN (tests LT JOIN (tests)) ON (Ticker)) ON (Ticker)`
4. the error should be reproduced

### Expected Behavior

The data should be returned from the query

### Environment

```markdown
- **QuestDB version**: Updated to 7 but webui still shows 6.6.1
- **OS**:  Microsoft Windows 10 Pro
- **Browser**: Google Chrome
```

### Additional context

image

### PR DESCRIPTION
**Pull request by jerrinot**
fixes #2976

The issue is best demonstrated with multiple levels of LT JOINS.
It goes like this:
1. LT JOIN physical tables. This causes no problem because physical tables support random access -> QuestDB uses light join. Light join does not copy columns from a slave cursor to a Map-backed record. However the JOIN result set no longer supports random access.
2. Join another table with the result of ☝️  and use a symbol column as a joining key. The slave cursor does not support random access -> QuestDB has to copy columns from the slave record to a map. But because we are joining on symbol keys, it has to convert symbols to a string and use the string as a key. 
3. The result of ☝️ has symbols from the slave table, used as keys, converted to strings :-(

The idea behind this fix: Join keys are present in both master and slave records. So we can still convert key symbols from the slave to strings to join with the master. But when a downstream code asks for a key column from the slave we can cheat and return the symbol ID from the master cursor. Because the symbol (string value) must be the same; otherwise, a join would not match. 

I call this: _"wrapping over from slave columns to master columns"_. Why? Columns in a join record as organized like this:
```
master_column0 | master_column1 | .... | master_columnN | slave_value_column0 | slave_value_column1 |... | slave_value_columnN | slave_key_column0 | slave_key_column1 | ... | slave_key_columnN
```
So when a downstream code asks for a symbol column which is after the slave key-value divide then we can "wrap-over" to master columns and fetch the symbol ID from a master record. Because the symbol has to have the same string value in both master and slave records. (assuming the slave record is non-null). And obviously, the join cursor has to return the right symbol table on `get/newSymbolTable()`

AN ALTERNATIVE SOLUTION:
An alternative could be to sink symbol IDs from keys into a map value. So when a JOIN key uses symbol then the Join would sink a string representation of the symbol to map key (for joining with a master record) and `int` representation of the symbol would also go to a map value. So when a downstream operator asks for a symbol ID which is used as a key the JOIN record would return the ID from a value portion of a map entry.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 65 / 73 (89.04%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|21|29|72.41%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 70 (91.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 70 (91.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 70 (91.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|

File: core/src/test/java/io/questdb/griffin/AsOfJoinTest.java [hunk 1]
```diff
@@ -292,6 +298,49 @@ public void testExplicitTimestampIsNotNecessaryWhenSpliceJoiningExplicitlyOrdere
         testExplicitTimestampIsNotNecessaryWhenJoining("splice join", null);
     }
 
+    @Test
+    public void testFullLtJoinDoesNotConvertSymbolKeyToString() throws SqlException {
+        compiler.setFullFatJoins(true);
+        compile("create table tab_a (sym_a symbol, ts_a timestamp, s_a string) timestamp(ts_a) partition by DAY");
+        compile("create table tab_b (sym_b symbol, ts_b timestamp, s_B string) timestamp(ts_b) partition by DAY");
+        compile("insert into tab_a values ('ABC', '2022-01-01T00:00:00.000000Z', 'foo')");
+        compile("insert into tab_b values ('ABC', '2021-01-01T00:00:00.000000Z', 'bar')");
```
Comment (ideoma) [inline core/src/test/java/io/questdb/griffin/AsOfJoinTest.java:307]: Here matching symbols are inserted first, they have the same id.
Can you change the test so that for example `tab_b` has another row with the symbol value `DCE` first?

**Review by ideoma — COMMENTED** [review]
Are there tests with LT join on symbol columns between different tables with unmatching symbol sets?

So that if table T1 has symbols 
1 = A
2 = B
3 = C

And table T2 has symbols
1 = C
2 = D
3 = A

and then join them

File: core/src/main/java/io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        if (col < split) {
+            return master.getSym(col);
+        }
+        int slaveCol = col - split;
+        if (isSlaveKeyColumn(slaveCol)) {
+            // key symbols are converted to strings before inserting into map.
+            // so we can read them as strings directly from the map.
+            return slave.getStr(slaveCol);
+        }
+        return slave.getSym(slaveCol);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java:78]: the outcome is the same, conditional met or not, bug?

File: core/src/main/java/io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public CharSequence getSymB(int col) {
+        if (col < split) {
+            return master.getSymB(col);
+        }
+        int slaveCol = col - split;
+        if (isSlaveKeyColumn(slaveCol)) {
+            return slave.getStrB(slaveCol);
+        } else {
+            return slave.getSymB(slaveCol);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java:90]: same, what is the magic here?

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/test/java/io/questdb/griffin/AsOfJoinTest.java (hunk 1 above)
Comment (jerrinot) [reply core/src/test/java/io/questdb/griffin/AsOfJoinTest.java:307]: good idea

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 70 (91.43%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|7|7|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 4]
```diff
@@ -556,21 +571,8 @@ private RecordCursorFactory createFullFatJoin(
...
-                        slaveTypes.add(ColumnType.STRING);
-                    } else {
-                        metadata.add(slaveAlias, m);
-                        slaveTypes.add(m.getType());
-                    }
+                    metadata.add(slaveAlias, m);
+                    slaveTypes.add(m.getType());
                     columnIndex.add(index);
                 }
             } else {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:578]: I don't see this `else` to be test covered with Symbol columns.

**Review by ideoma — COMMENTED** [review]
I'm still not sure I can fully understand the mechanics.
Perhaps @bziobrowski you can have a look here too?

File: core/src/main/java/io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        if (columnIndex < masterSlaveSplit) {
+            return masterCursor.getSymbolTable(columnIndex);
+        }
+        int slaveCol = columnIndex - masterSlaveSplit;
+        if (slaveCol >= slaveWrappedOverMaster) {
+            slaveCol -= slaveWrappedOverMaster;
+            int masterCol = masterTableKeyColumns.getColumnIndexFactored(slaveCol);
+            return masterCursor.getSymbolTable(masterCol);
+        }
+        slaveCol = slaveColumnIndex.getQuick(slaveCol);
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java:67]: yes, this path is taken happen when a symbol column from the slave table is not used a join key yet it's projected to the JoinRecord. 

the path is tested from many tests
image

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java (hunk 4 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:578]: good one, let me have a look!

Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:578]: hm, you are right. 
I ran all tests in the `io.questdb.test.griffin` package and the `else` branch was not executed *at all* 😱

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java [hunk 6]
```diff
@@ -798,6 +877,108 @@ public void testLtJoinSequenceGapOnKey() throws Exception {
...
+    public void testNestedASOF_keySymbol() throws Exception {
+        assertMemoryLeak(() -> {
+            compiler.setFullFatJoins(true);
+            compile("CREATE TABLE 'tests' (\n" +
+                    "  Ticker SYMBOL capacity 256 CACHE,\n" +
+                    "  ts timestamp\n" +
+                    ") timestamp (ts) PARTITION BY MONTH");
+            compile("insert into tests VALUES ('AAPL', '2000'),('AAPL', '2001'),('AAPL', '2002'),('AAPL', '2003'),('AAPL', '2004'),('AAPL', '2005')");
+            compile("insert into tests VALUES ('QSTDB', '2003'),('QSTDB', '2004'),('QSTDB', '2005'),('QSTDB', '2006'),('QSTDB', '2007'),('QSTDB', '2008')");
+
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:890]: It would be good  to format queries a bit .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 101 (92.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

**Comment by jerrinot** [comment]
@ideoma can you please elaborate this:
> One more question, what if the Master table does not have a NULL value in the symbol columns but the NULL column value is the result of the ASOF join? Can you please add tests with such asof join and IS NULL filter in outer select?

Do you mean joining on a symbol key where a slave cursor has no matching record?

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 6 above)
Comment (jerrinot) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:904]: is the current version better?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 101 (92.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java [hunk 7]
```diff
@@ -145,10 +146,10 @@ public AsOfJoinRecordCursor(
                 Record nullRecord,
                 int masterTimestampIndex,
                 int slaveTimestampIndex,
-                RecordValueSink valueSink
-        ) {
-            super(columnSplit);
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java:150]: Could you fix formatting here ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 101 (92.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 6 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:929]: Yes 👍

File: core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java (hunk 7 above)
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java:150]: done

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java [hunk 8]
```diff
@@ -798,6 +902,174 @@ public void testLtJoinSequenceGapOnKey() throws Exception {
...
+                    "QSTDB\t2004-01-01T00:00:00.000000Z\tQSTDB\t2003-01-01T00:00:00.000000Z\t\t\t\t\n" +
+                    "AAPL\t2004-01-01T00:00:00.000000Z\tAAPL\t2003-01-01T00:00:00.000000Z\tAAPL\t2002-01-01T00:00:00.000000Z\tAAPL\t2001-01-01T00:00:00.000000Z\n" +
+                    "AAPL\t2005-01-01T00:00:00.000000Z\tAAPL\t2004-01-01T00:00:00.000000Z\tAAPL\t2003-01-01T00:00:00.000000Z\tAAPL\t2002-01-01T00:00:00.000000Z\n" +
+                    "QSTDB\t2005-01-01T00:00:00.000000Z\tQSTDB\t2004-01-01T00:00:00.000000Z\tQSTDB\t2003-01-01T00:00:00.000000Z\t\t\n" +
+                    "QSTDB\t2006-01-01T00:00:00.000000Z\tQSTDB\t2005-01-01T00:00:00.000000Z\tQSTDB\t2004-01-01T00:00:00.000000Z\tQSTDB\t2003-01-01T00:00:00.000000Z\n" +
+                    "QSTDB\t2007-01-01T00:00:00.000000Z\tQSTDB\t2006-01-01T00:00:00.000000Z\tQSTDB\t2005-01-01T00:00:00.000000Z\tQSTDB\t2004-01-01T00:00:00.000000Z\n" +
+                    "QSTDB\t2008-01-01T00:00:00.000000Z\tQSTDB\t2007-01-01T00:00:00.000000Z\tQSTDB\t2006-01-01T00:00:00.000000Z\tQSTDB\t2005-01-01T00:00:00.000000Z\n";
+            assertQuery(expected, query, "ts", false, true);
+        });
+    }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1013]: I reckon it'd be good to add test(s) with :
- more than  one join key 
- synthetic symbol column in master as join key (e.g. `select rnd_symbol('a', 'b')  ticker, ts from tests ` )

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java [hunk 9]
```diff
@@ -291,6 +297,80 @@ public void testExplicitTimestampIsNotNecessaryWhenSpliceJoiningExplicitlyOrdere
         testExplicitTimestampIsNotNecessaryWhenJoining("splice join", null);
     }
```
Comment (bziobrowski) [inline core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:299]: Since PR affects lt and asof shouldn't there be a similar test case for asof join ?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 99 (91.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 8 above)
Comment (jerrinot) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1067]: added `testLtJoinOnCompositeSymbolKey()`. 
will add a test with a synthetic join column in a master, that's a good idea!

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 9 above)
Comment (jerrinot) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:299]: good one

Comment (jerrinot) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:299]: done

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 99 (91.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 99 (91.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByUtils.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/SymbolWrapOverJoinRecord.java|20|26|76.92%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|25|26|96.15%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AbstractSymbolWrapOverCursor.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/JoinRecordMetadata.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/LtJoinRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractRecordMetadata.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GenericRecordMetadata.java|8|8|100.00%|

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 8 above)
Comment (jerrinot) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1119]: done

**Comment by jerrinot** [comment]
a failure on windows-other due to the https://github.com/questdb/questdb/issues/2941
re-running

**Review by bziobrowski — APPROVED** [review]
Looks good now 👍
